homeassistant/components/qbus/config_flow.py

```diff
@@ -0,0 +1,159 @@
+"""Config flow for Qbus."""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, Any
+
+from qbusmqttapi.discovery import QbusMqttDevice
+from qbusmqttapi.factory import QbusMqttMessageFactory, QbusMqttTopicFactory
+
+from homeassistant import config_entries
+from homeassistant.components.mqtt import client as mqtt
+from homeassistant.config_entries import ConfigFlowResult
+from homeassistant.helpers.service_info.mqtt import MqttServiceInfo
+
+from .const import CONF_ID, CONF_SERIAL, DOMAIN
+from .qbus import QbusConfigContainer
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class QbusFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle Qbus config flow."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize."""
+        self._message_factory = QbusMqttMessageFactory()
+        self._topic_factory = QbusMqttTopicFactory()
+
+        self._gateway_topic = self._topic_factory.get_gateway_state_topic()
+        self._config_topic = self._topic_factory.get_config_topic()
+        self._controller_topic = self._topic_factory.get_device_state_topic("+")
+
+        self._device: QbusMqttDevice | None = None
+
+    async def async_step_mqtt(
+        self, discovery_info: MqttServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by MQTT discovery."""
+        _LOGGER.debug("Running mqtt discovery for topic %s", discovery_info.topic)
+
+        # Abort if the payload is empty
+        if not discovery_info.payload:
+            _LOGGER.debug("Payload empty")
+            return self.async_abort(reason="invalid_discovery_info")
+
+        match discovery_info.subscribed_topic:
+            case self._gateway_topic:
+                return await self._async_handle_gateway_topic(discovery_info)
+
+            case self._config_topic:
+                return await self._async_handle_config_topic(discovery_info)
+
+            case self._controller_topic:
+                return await self._async_handle_controller_topic(discovery_info)
```

Comment:
I am not sure if we should have this complexity in the discovery. I'll have to get someone else's opinion.
It would be great to move this out of discovery in just have a single method that does the 3 "setup" steps

Reply:
You mean the `match` block? It acts directly on the incoming `discovery_info` parameter and with the `match` you have a clear view on what's exactly happening in the discovery. I can move the `match` block (and the `async_abort` following it) to a private method so it becomes a single line in the discovery method, but I don't see the benefit to be honest. But let me know what you want, and I'll adapt if necessary.

Those 3 methods cannot be handled in a single method, it'll always need to be a match/case. The Qbus MQTT gateway flow is as follows:

1. When it comes online or goes offline, it will publish to the gateway topic. In `_async_handle_gateway_topic` we check if the state is "online", and if so we request the latest config by publishing to a topic.
2. The Qbus MQTT gateway responds in another topic with the topic and now we land in the second case: `_async_handle_config_topic`. This methods inspects the config for devices and for each devices it finds, it will request the current state for that device.
3. The Qbus MQTT gateway responds with the state of each requested device, this is handled in `_async_handle_device_topic`. Here we present the user with the discovered devices.

Reply:
Thanks for the overview!